src/plugins/vis_types/vega/public/vega_view/vega_base_view.js

```diff
@@ -44,6 +45,47 @@ for (const funcName of Object.keys(vegaFunctions)) {
 
 const bypassToken = Symbol();
 
+function normalizeDate(date) {
+  let normalizedDate = '';
+  if (typeof date === 'number') {
+    normalizedDate = date;
+  } else if (date instanceof Date) {
+    normalizedDate = date;
+  } else {
+    normalizedDate = String(normalizeObject(date));
+  }
+  return normalizedDate;
+}
+
+// recursive function to check a nested object for a function property
+function checkObjectForFunctionProperty(object) {
+  if (object === null || object === undefined) {
+    return false;
+  }
+  if (typeof object === 'function') {
+    return true;
+  }
+  const hasFnProperty = Object.keys(object).some((key) => {
+    if (typeof object[key] === 'function') {
+      return checkObjectForFunctionProperty(object[key]);
+    }
+  });
+  return hasFnProperty;
+}
+
+function normalizeObject(object) {
+  if (checkObjectForFunctionProperty(object)) {
+    throw new Error('a function cannot be used as a property name');
+  }
+  const normalizedObject = JSON.parse(JSON.stringify(object));
```

Comment:
`JSON.stringify` will also remove functions. Is there a reason why we want to throw in case of a function instead of just silently remove it?

If we do want to throw, a more efficient way would be to use the `replacer` argument to `JSON.stringify`. This way we don't have to walk the object twice (once in `checkObjectForFunctionProperty` and once by `JSON.stringify`):

```suggestion
function throwOnFunction(key, value) {
  if (typeof value === 'function') {
    throw new Error('a function cannot be used as a property name');
  }
  return value;
}

function normalizeObject(object) {
  const normalizedObject = JSON.parse(JSON.stringify(object, throwOnFunction));
```

Reply:
I had a hunch that using a `replacer` might not be that performant after all since it probably will have to call back and forth between JS land and C/C++ land. So I did [a quick and dirty perf test](https://gist.github.com/watson/4eb141686a7f339a1d21039a1a0d0650):

no function check: 6.031s
with replacer that throws: 12.094s
with replacer that do not throw: 11.852s
with separate loop: 6.931s

So while I think my use of a `replacer` is a lot more readable, it's less performant (at least in Node.js/V8 - other JavaScript engines might vary).

It's also worth considering that creating your own object-walker algorithm is hard to get right. So using the `replacer` function is a bit more "secure"/"safe".

Reply:
> Is there a reason why we want to throw in case of a function instead of just silently remove it?

Good question, I'm a little on the fence. I can't think of any good use case to have a function in that object, but so far we didn't throw on them (and as these objects are queries sent to Elasticsearch they are turned into a JSON string anyway in a later stage), so silently removing would stick closer to the existing behavior for the off chance someone had a function in there by accident which was silently removed and now we start throwing. However, maybe it's OK we start throwing and leaving a note for users running into this in the release notes.

Reply:
++ to the `replacer`-based version even though it's a bit less performant, it doesn't seem critical for this use case. Regarding throwing or silently removing functions, throwing an error might help the users to debug the problem if they define functions by mistake (assuming the error is displayed to the user).

Reply:
In summary, I'm happy with a simpler solution as long as it still catches all functions before executing them (even a `toJSON` function itself)

Reply:
However, actually testing this I noticed that it doesn't work:
```
function throwOnFunction(key, value) {
  if (typeof value === 'function') {
    throw new Error('a function cannot be used as a property name');
  }
  return value;
}

JSON.parse(JSON.stringify({ toJSON: () => alert('gotcha')}, throwOnFunction))
```

Will execute the `toJSON` function.

Reply:
![giphy](https://user-images.githubusercontent.com/10602/172578843-8a36ba83-33e7-4700-b75d-be4acd7c1fc3.gif)

Thanks for being vigilant @flash1293

Reply:
He is! I addressed all the comments! Thank you both for your support on this issue!